[BUG] `exec`/`npx`: npm v8.8+/9+ break with a self-reference

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

In a package `foo` with a bin, run `npx foo`, and you'll see `sh: foo: command not found`.

### Expected Behavior

In a package `foo` with a bin, run `npx foo`, and the bin is invoked.

### Steps To Reproduce

1. Clone any repo, presumably, but in this case, https://github.com/ljharb/publishers
2. Run `npx npm@8.7 exec publishers`, see that it invokes the binary properly.
3. Run `npx npm@8.8 exec publishers`, see that it errors.

### Environment

- npm: <= 8.7, >= 8.8
- Node.js: 19

Running the original steps to reproduce, I get a different error with `9.1.1`:

```
❯ npx --version
9.1.1

❯ pwd
/Users/lukekarrys/projects/ljharb/publishers

❯ npx publishers
node:internal/modules/cjs/loader:998
  throw err;
  ^

Error: Cannot find module 'json-colorizer'
Require stack:
- /Users/lukekarrys/projects/ljharb/publishers/bin/publishers
    at Module._resolveFilename (node:internal/modules/cjs/loader:995:15)
    at Module._load (node:internal/modules/cjs/loader:841:27)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/lukekarrys/projects/ljharb/publishers/bin/publishers:5:18)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/Users/lukekarrys/projects/ljharb/publishers/bin/publishers' ]
}

Node.js v18.12.1
```

--- a/workspaces/libnpmexec/lib/index.js
+++ b/workspaces/libnpmexec/lib/index.js
@@ -110,7 +110,7 @@ const exec = async (opts) => {
     return run()
   }
 
-  const needPackageCommandSwap = (args.length > 0) && (packages.length === 0)
+  let needPackageCommandSwap = (args.length > 0) && (packages.length === 0)
   // If they asked for a command w/o specifying a package, see if there is a
   // bin that directly matches that name:
   // - in the local package itself
@@ -126,9 +126,11 @@ const exec = async (opts) => {
     if (localManifest?.bin?.[args[0]]) {
       // we have to install the local package into the npx cache so that its
       // bin links get set up
+      flatOptions.installLinks = false
+      // args[0] will exist when the package is installed
       packages.push(path)
       yes = true
-      flatOptions.installLinks = false
+      needPackageCommandSwap = false
     } else {
       const dir = dirname(dirname(localBin))
       const localBinPath = await localFileExists(dir, args[0], '/')

--- a/workspaces/libnpmexec/test/index.js
+++ b/workspaces/libnpmexec/test/index.js
@@ -28,6 +28,7 @@ t.test('bin in local pkg', async t => {
   const pkg = {
     name: '@npmcli/local-pkg-bin-test',
     bin: {
+      b: 'echo this is something else',
       a: 'local-bin-test.js',
     },
   }

